src/utils/content_analysis.py

```
@@ -0,0 +1,513 @@
+        
+        logger.debug("Extracted %d valid code blocks from %d total found", len(valid_blocks), len(code_blocks))
+        return valid_blocks
+        
+    except Exception as e:
+        logger.exception("Error in extract_code_blocks: %s", e)
+        return []
+
+
+async def ingest_pages_to_neo4j(
+    pages_data: list[dict[str, Any]],
+    report_progress_func: Optional[callable] = None
```

Ruff (0.12.2)

Use `X | None` for type annotations

Convert to `X | None`

(UP045)